pkgs/top-level/stage.nix

```
@@ -182,6 +182,14 @@ let
                               nixpkgsFun { inherit crossSystem; })
                               lib.systems.examples;
 
+    # The logical inverse of pkgsCross, so that pkgs.pkgsCross.*.pkgsNative == pkgs,
+    # (and pkgs.pkgsNative == pkgs).
+    # This way, you can cross compile most things; but replace that one specific
+    # package that fails to be cross compiled with the native one.
+    pkgsNative =
+      nixpkgsFun { localSystem = stdenv.hostPlatform; crossSystem = null; }
```

```suggestion
    pkgsNative = nixpkgsFun {
      overlays = [
        (self': super': {
          pkgsNative = super';
        })
      ] ++ overlays;
      localSystem = stdenv.hostPlatform; crossSystem = null;
    }
```

the `otherPackageSets` other than `pkgsCross` do this

---

Dunno if correct here

```
nix-repl> pkgsCross.aarch64-multiplatform.pkgsNative.bash
«derivation /nix/store/xk7bw18arb5k9kkajiflrfh9jgv6nxmw-bash-5.2-p15.drv»

nix-repl> pkgsCross.aarch64-multiplatform.pkgsNative.pkgsNative.bash
«derivation /nix/store/3qczm1q9j3553pxi8lvkwj8vqq1zx5lw-bash-5.2-p15.drv»
```

```
nix-repl> pkgsLLVM.bash
«derivation /nix/store/c8fgbkk89y8019nbyqps8y48r5nk9zr6-bash-5.2-p15-x86_64-unknown-linux-gnu.drv»

nix-repl> pkgsLLVM.pkgsLLVM.bash
«derivation /nix/store/c8fgbkk89y8019nbyqps8y48r5nk9zr6-bash-5.2-p15-x86_64-unknown-linux-gnu.drv»
```

---

I still wonder why the hash is not the same

```
nix-repl> pkgsCross.aarch64-multiplatform.pkgsNative.bash
«derivation /nix/store/xk7bw18arb5k9kkajiflrfh9jgv6nxmw-bash-5.2-p15.drv»

nix-repl> pkgsCross.aarch64-multiplatform.pkgsNative.pkgsNative.bash
«derivation /nix/store/3qczm1q9j3553pxi8lvkwj8vqq1zx5lw-bash-5.2-p15.drv»
```

---

And if it should be `= self'` and if overlays work properly

---

I'm getting the correct hashes?
```nix
# build == host == aarch64-linux
nix-repl> pkgs = import ./. { localSystem.config = "aarch64-unknown-linux-gnu"; } 

nix-repl> pkgs.hello
«derivation /nix/store/440grns9r25bhcs6jcw16lr6glszgpw4-hello-2.12.1.drv»

nix-repl> pkgs.pkgsNative.hello
«derivation /nix/store/440grns9r25bhcs6jcw16lr6glszgpw4-hello-2.12.1.drv»

nix-repl> pkgs.pkgsNative.pkgsNative.hello
«derivation /nix/store/440grns9r25bhcs6jcw16lr6glszgpw4-hello-2.12.1.drv»

# build == x86_64-linux != host == aarch64-linux
nix-repl> pkgs = import ./. { }

nix-repl> pkgs.pkgsCross.aarch64-multiplatform.pkgsNative.hello
«derivation /nix/store/440grns9r25bhcs6jcw16lr6glszgpw4-hello-2.12.1.drv»

nix-repl> pkgs.pkgsCross.aarch64-multiplatform.pkgsNative.pkgsNative.hello
«derivation /nix/store/440grns9r25bhcs6jcw16lr6glszgpw4-hello-2.12.1.drv»
```

---

Oh, it works for hello but for bash it gives different hashes? I get the same as you with bash

---

When using `self'` it works.

```nix
nix-repl> pkgs = import ./. { }                                           

nix-repl> pkgs.pkgsCross.aarch64-multiplatform.pkgsNative.bash
«derivation /nix/store/xk7bw18arb5k9kkajiflrfh9jgv6nxmw-bash-5.2-p15.drv»

nix-repl> pkgs.pkgsCross.aarch64-multiplatform.pkgsNative.pkgsNative.bash
«derivation /nix/store/xk7bw18arb5k9kkajiflrfh9jgv6nxmw-bash-5.2-p15.drv»
```

And I think using super' is wrong, since as you said otherwise it won't apply any overlays.

---

Tried it out and using `self'` here has the same infinite recursion issue, and the evaluator indeed consumes massive amounts of memory with the original function that I did. What surprises me is that using `super'` does apply overlays properly. But of course we're back at the different bash hashes.

---

Ok, started comparing things and found this

```nix
nix-repl> pkgs.pkgsCross.aarch64-multiplatform.pkgsNative.pkgsNative.bash.builder
"/nix/store/zkp0y089k5a7z4iwpapfrifiqa3i8y67-bash-5.2-p15/bin/bash"

nix-repl> pkgs.pkgsCross.aarch64-multiplatform.pkgsNative.bash.builder            
"/nix/store/frjyqvav0a1jky8g6sr70vdf71hjkcys-bootstrap-tools/bin/bash"
```

---

Doing this little work around fixes the issue

```nix
pkgsNative = super'.pkgsNative or super';
```

```nix
nix-repl> pkgs = import ./. { }                                                    

nix-repl> pkgs.pkgsCross.aarch64-multiplatform.pkgsNative.pkgsNative.bash          
«derivation /nix/store/xk7bw18arb5k9kkajiflrfh9jgv6nxmw-bash-5.2-p15.drv»

nix-repl> pkgs.pkgsCross.aarch64-multiplatform.pkgsNative.bash            
«derivation /nix/store/xk7bw18arb5k9kkajiflrfh9jgv6nxmw-bash-5.2-p15.drv»

nix-repl> pkgs = import ./. { localSystem.config = "aarch64-unknown-linux-gnu"; }          

nix-repl> pkgs.bash                                                               
«derivation /nix/store/xk7bw18arb5k9kkajiflrfh9jgv6nxmw-bash-5.2-p15.drv»

nix-repl> pkgs.pkgsNative.bash
«derivation /nix/store/xk7bw18arb5k9kkajiflrfh9jgv6nxmw-bash-5.2-p15.drv»
```

Though it somehow does not look right to me.